### ISSUE TITLE
Adding static methods/vars to a stampit constructor?

### ISSUE THREAD

**Issue by DylanPiercey** [ericelliott/stampit#64]
Is there any convenience method to add statics methods to an object constructor or does this have to be handled outside of stampit?

**Comment by troutowicz** [comment]
I don't think so. The way I have been doing this is:

```js
...
// create stamp
...
export default stampit.extend(someStamp, {
  // static methods/vars here
});
```

**Comment by ericelliott** [comment]
@troutowicz nailed it. That's how I do it. :+1:

**Comment by ericelliott** [comment]
A `.static()` convenience method would be a nice addition to the 1.x line. We should also pull it into 2.x, of course. Do either of you want to do a PR?

@koresar, you want to lead this? If @DylanPiercey or @troutowicz want to do a PR, do you have time to review & merge?

**Comment by troutowicz** [comment]
@ericelliott I can work on the PR :)

**Comment by ericelliott** [comment]
https://github.com/ericelliott/stampit/pull/65

**Comment by ericelliott** [comment]
@troutowicz I've added you as a collaborator and assigned the issue to you. Feel free to reassign to @koresar when you're ready for code review.

**Comment by koresar** [comment]
Hey @troutowicz, thanks for the PR.

I like the idea of having statics for stamp!

However, I probably should inform that we are going to have "refs" feature (see #60). Basically, these are object references added to each object instance (not stamp). Statics in the object so to say. May I wonder why you need statics for stamps but not object instances? I'd like to find how many people would need this feature.

Also, stampit main objective are objects, not stamps. I doubt people would use this feature extensively, if use at all. I believe modules (like stampit) should do single job and do it well.

Moreover, I feel like introducing a whole new feature which saves people 15 characters
`stampit.extend(someStamp, { ... });` is not worth the hassle around it. We would need to support it preferably forever.

Please, explain me where I'm confusing things. I believe you have stong need in this feature. Some more dicussion is really necessary.

Thanks @DylanPiercey for the idea! Cheers

**Comment by troutowicz** [comment]
Hi @koresar.

So `refs` is just the new `state`, and `state` binds to the stamp instance. That feature has it's uses, but some libraries, like [React](https://github.com/facebook/react), look at the constructor object for [certain properties](https://facebook.github.io/react/blog/2015/01/27/react-v0.13.0-beta-1.html#es6-classes). And this is where a declarative `static` convenience method would be super cool.

At the end of most React components, you will find several `objects` being assigned to the `class` declared earlier in the component.

Example from my [project](https://github.com/troutowicz/geoshare)'s auth button:
```js
...
// class/stamp declared
...
// add several objects to class/stamp
AuthButton.contextTypes = {
  muiTheme: React.PropTypes.object
};

AuthButton.propTypes = {
  label: React.PropTypes.string,
  onTouchTap: React.PropTypes.func,
  style: React.PropTypes.object
};

AuthButton.defaultProps = {
  label: 'Login',
  style: {}
};
...
```
This imperative step of designing a component could be done away with if we could declare these objects inside our stamp. I realize that these property assignments could be made more declarative using a helper function like `mixIn/extend/Object.assign`, but that still doesn't do away with the unnecessary extra step.

I should also mention that class [property initializers](https://esdiscuss.org/topic/es7-property-initializers), like `static`, are being discussed for ES7.

Does that clear up the need for this feature?

**Comment by koresar** [comment]
Oh, that explains it all. Thank you so much.

I saw the example of unwanted syntax in #65. Is the following the proper usage of the proposed API?
```js
var AuthButton = stampit({
// methods
}, {
// state
},
function () {
// enclose
}, {
// static
});
```

If yes, then we have stampit v2.0 incompatibility issue. If you take a closer look at #60 changes (which I'm going to finish tomorrow) the `stampit()` function receives a fourth parameter for the deep cloned state. Thus, your PR is heavily conflicting with #60.

If we add the fifth parameter to the `stampit()` then we are introducing heavy code smell. :) Even fourth parameter is a code smell too. Someday someone would propose sixth, and so on. :)

I believe stampit v2.0 should improve its core API - the `stampit()` main function.
```js
var stamp = stampit({
  refs: {},
  methods: {},
  props: {},
  enclose: function () {}, // or an array of functions
  static: {}
});
```

The implementation is trivial (ducktyping). Would that suffice your needs, guys?

**Comment by troutowicz** [comment]
We could improve the core API like you describe, or just not support the passing of a `static` arg. Although we probably want to support all use cases.

**Comment by ericelliott** [comment]
Oh crap, I edited @koresar's post instead of creating my own. It was an accident, and I am a clown.

Here's my proper reply. I don't know how to undo my crappy edit:

> However, I probably should inform that we are going to have "refs" feature (see #60). Basically, these are object references added to each object instance (not stamp). Statics in the object so to say. May I wonder why you need statics for stamps but not object instances? I'd like to find how many people would need this feature.

There are already several ways to set properties on instances, and we definitely don't need another. A static is a property set on the factory itself (or the constructor in classical OO, or the class itself in some other OO languages).

> Also, stampit main objective are objects, not stamps. I doubt people would use this feature extensively, if use at all. I believe modules (like stampit) should do single job and do it well.

I disagree completely, not least because React components have some properties that are preferred as statics, including `propTypes`. Type specification is a really good use-case for factories, in fact, because they tell you exactly how to use the stamp, including the types of all the parameters and the return value. Among other things, this can be used to provide type hints in IDEs, or even automatic detection of which object types are composable in graphical dataflow systems like [NoFlo](http://noflojs.org/).

> Moreover, I feel like introducing a whole new feature which saves people 15 characters
`stampit.extend(someStamp, { ... });` is not worth the hassle around it. As we would need to support it preferably forever.

Because a lot of people will likely end up using Stampit to create composable React components, and will likely create many components per day, I have to disagree again.

Thanks for pushing back. We definitely do need to justify new features to avoid bloating Stampit and avoid including features that are out of scope, but in this case, I think we need to do it. =)

**Comment by ericelliott** [comment]
```js
var stamp = stampit({
  refs: {},
  methods: {},
  props: {},
  enclose: function () {}, // or an array of functions
  static: {}
});
```

This is great for the 2.x line.

For the 1.x line, only include the method form, not the parameter form. I don't want people to start using 5 params, and we can't make a breaking change to the 1.x line.

**Comment by troutowicz** [comment]
:+1:

**Comment by ericelliott** [comment]
I just thought of another great use for a `propTypes` static property on stamps: Static analysis could automatically detect if composing stamps would cause a property name conflict, and warn if they do. We could create plugins for popular IDE's, and it's possible the React team might create custom tooling for Babel that stamps could automatically benefit from.

**Comment by troutowicz** [comment]
That would be awesome.

**Comment by koresar** [comment]
So, the outcome of this issue are two issues.

Implemented in stampit v1 under #76.
The v2 implementation to follow under #74.

Closing this issue.

Big thanks guys!
